From 4b4f63a41d4ee33a4b3c048bb4281051acbad106 Mon Sep 17 00:00:00 2001
From: Jordan Williams <jordan@jwillikers.com>
Date: Tue, 7 Nov 2023 07:38:42 -0600
Subject: [PATCH] Use find_package and pkg_check_modules to find more
 dependencies

This commit enhances the use of the FindOpenGL CMake module.
This requires CMake version 3.10 for the OpenGL::EGL imported target.
CMake 3.11 and later enable CMake policy CMP0072.
This prefers the GLVND libraries when available.

Finds the Wayland and xkbcommon dependencies with pkg_check_modules.
This works with the pkg-config files provided by the upstream projects.
---
 CMakeLists.txt | 53 +++++++++++++++++++++++++++++++++++++++++++-------
 1 file changed, 46 insertions(+), 7 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index ac8c8df3..f97f14ff 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,5 +1,9 @@
-CMAKE_MINIMUM_REQUIRED(VERSION 3.0.0 FATAL_ERROR)
-PROJECT(freeglut)
+CMAKE_MINIMUM_REQUIRED(VERSION 3.1 FATAL_ERROR)
+PROJECT(freeglut LANGUAGES C)
+
+if (POLICY CMP0072)
+  cmake_policy(SET CMP0072 NEW)
+endif()
 
 # for multiarch LIBDIR support (requires cmake>=2.8.8)
 INCLUDE(GNUInstallDirs)
@@ -261,17 +265,52 @@ ENDIF()
 # GLES1 and GLES2 libraries are compatible and can be co-linked.
 IF(FREEGLUT_GLES)
     LIST(APPEND PUBLIC_DEFINITIONS -DFREEGLUT_GLES)
-    LIST(APPEND LIBS GLESv2 GLESv1_CM EGL)
+    if(NOT CMAKE_VERSION VERSION_LESS "3.27")
+      FIND_PACKAGE(OpenGL REQUIRED COMPONENTS EGL GLES2 OpenGL)
+      LIST(APPEND LIBS GLESv1_CM OpenGL::EGL OpenGL::GLES2 OpenGL::OpenGL)
+    elseif(NOT CMAKE_VERSION VERSION_LESS "3.10")
+      FIND_PACKAGE(OpenGL REQUIRED COMPONENTS EGL OpenGL)
+      LIST(APPEND LIBS GLESv2 GLESv1_CM OpenGL::EGL OpenGL::OpenGL)
+    else()
+      FIND_PACKAGE(OpenGL REQUIRED)
+      LIST(APPEND LIBS EGL GLESv2 GLESv1_CM)
+    endif()
 ELSE()
-  FIND_PACKAGE(OpenGL REQUIRED)
-  LIST(APPEND LIBS ${OPENGL_gl_LIBRARY})
-  INCLUDE_DIRECTORIES(${OPENGL_INCLUDE_DIR})
+    if(NOT CMAKE_VERSION VERSION_LESS "3.10")
+      FIND_PACKAGE(OpenGL REQUIRED COMPONENTS OpenGL)
+      LIST(APPEND LIBS OpenGL::GL)
+    else()
+      FIND_PACKAGE(OpenGL REQUIRED)
+      LIST(APPEND LIBS ${OPENGL_gl_LIBRARY})
+      INCLUDE_DIRECTORIES(${OPENGL_INCLUDE_DIR})
+    endif()
 ENDIF()
 
 # For Wayland: compile with -DFREEGLUT_WAYLAND and pull EGL
 IF(FREEGLUT_WAYLAND)
   ADD_DEFINITIONS(-DFREEGLUT_WAYLAND)
-  LIST(APPEND LIBS wayland-client wayland-cursor wayland-egl EGL xkbcommon)
+  INCLUDE(FindPkgConfig)
+  if(NOT CMAKE_VERSION VERSION_LESS "3.10")
+    FIND_PACKAGE(OpenGL REQUIRED COMPONENTS EGL OpenGL)
+    LIST(APPEND LIBS OpenGL::EGL OpenGL::OpenGL)
+  else()
+    FIND_PACKAGE(OpenGL REQUIRED)
+    LIST(APPEND LIBS EGL)
+  endif()
+  if(NOT CMAKE_VERSION VERSION_LESS "3.6")
+    PKG_CHECK_MODULES(wayland-client REQUIRED IMPORTED_TARGET wayland-client)
+    PKG_CHECK_MODULES(wayland-cursor REQUIRED IMPORTED_TARGET wayland-cursor)
+    PKG_CHECK_MODULES(wayland-egl REQUIRED IMPORTED_TARGET wayland-egl)
+    PKG_CHECK_MODULES(xkbcommon REQUIRED IMPORTED_TARGET xkbcommon)
+    LIST(APPEND LIBS PkgConfig::wayland-client PkgConfig::wayland-cursor PkgConfig::wayland-egl PkgConfig::xkbcommon)
+  else()
+    PKG_CHECK_MODULES(wayland-client REQUIRED)
+    PKG_CHECK_MODULES(wayland-cursor REQUIRED)
+    PKG_CHECK_MODULES(wayland-egl REQUIRED)
+    PKG_CHECK_MODULES(xkbcommon REQUIRED)
+    LIST(APPEND LIBS ${wayland-client_LINK_LIBRARIES} ${wayland-cursor_LINK_LIBRARIES} ${wayland-egl_LINK_LIBRARIES} ${xkbcommon_LINK_LIBRARIES})
+    INCLUDE_DIRECTORIES(${wayland-client_INCLUDE_DIRS} ${wayland-cursor_INCLUDE_DIRS} ${wayland-egl_INCLUDE_DIRS} ${xkbcommon_INCLUDE_DIRS})
+  endif()
 ENDIF()
 
 # lib m for math, not needed on windows
-- 
2.41.0

